src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -657,8 +673,134 @@ public override Task WriteAsync(Byte[] array, int offset, int count, Cancellatio
 
         public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
         {
-            return StreamHelpers.ArrayPoolCopyToAsync(this, destination, bufferSize, cancellationToken: cancellationToken);
+            // Validation as base CopyToAsync would do
+            StreamHelpers.ValidateCopyToArgs(this, destination, bufferSize);
+
+            // Validation as ReadAsync would do
+            EnsureDecompressionMode();
+            if (_asyncOperations != 0) throw new InvalidOperationException(SR.InvalidBeginCall);
+            EnsureNotDisposed();
+
+            // Early check for cancellation
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<int>(cancellationToken);
+            }
+
+            // The caller asked for a specific bufferSize, but we're going to end up potentially using two buffers:
+            // a buffer for storing the inflated output and passing along to the destination stream, and whatever
+            // buffer the source stream creates in its CopyToAsync implementation.  As such, we halve the buffer size,
+            // to split it across the two.
+            bufferSize = Math.Max(1, bufferSize / 2);
+
+            // Do the copy
+            return new CopyToAsyncStream(this, destination, ArrayPool<byte>.Shared.Rent(bufferSize), cancellationToken).CopyFromSourceToDestination();
```

Comment:
Very minor nit:  seems like it would be better to rent the ArrayPool in CopyToAsyncStream; that way it's more cleanly encapsulated if we ever wanted to change to a different buffer pool implementation.